Mathlib/LinearAlgebra/RootSystem/Finite/Nondegenerate.lean

```diff
@@ -0,0 +1,167 @@
+    (smul_right_injective N (Ne.symm (ne_of_lt P.prod_rootForm_root_self_pos)))
+    (mem_center_iff.mp (by simp)) fun _ hx => ?_
+  obtain ⟨c, hc⟩ := (mem_span_range_iff_exists_fun R).mp hx
+  rw [← hc, Finset.smul_sum]
+  simp_rw [smul_smul, mul_comm, ← smul_smul]
+  exact Submodule.sum_smul_mem
+    (LinearMap.range (P.Polarization.domRestrict (span R (range P.root)))) c
+    (fun c _ ↦ prod_rootForm_smul_coroot_in_range_domRestrict P c)
+
+lemma rank_coPolarization_domRestrict_eq_rank_span_root :
+    LinearMap.rank (P.CoPolarization.domRestrict (span R (range P.coroot))) =
+      Module.rank R (span R (range P.root)) :=
+  P.flip.rank_polarization_domRestrict_eq_rank_span_coroot
+
+lemma finrank_span_root_le :
+    Module.finrank R (span R (range P.coroot)) ≤ Module.finrank R (span R (range P.root)) := by
+  have h := lift_rank_map_le P.Polarization (span R (range P.root))
+  refine finrank_le_finrank_of_rank_le_rank (le_of_eq_of_le ?_ h) (span_root_finite_rank P)
+  rw [← rank_polarization_domRestrict_eq_rank_span_coroot, ← LinearMap.range_domRestrict]
+
+lemma finrank_span_root_eq :
+    Module.finrank R (span R (range P.coroot)) = Module.finrank R (span R (range P.root)) :=
+  Nat.le_antisymm (finrank_span_root_le P) (by simpa using finrank_span_root_le P.flip)
+
+lemma polarization_domRestrict_kernel_rank_zero : Module.rank R (LinearMap.ker
+    (LinearMap.domRestrict P.Polarization (span R (range P.root)))) = 0 := by
+  let n := finrank R (span R (range P.root))
+  have hm : Module.rank R (span R (range P.root)) = n := by rw [finrank_eq_rank_span_root]
+  have hn : Module.rank R (LinearMap.range
+      (LinearMap.domRestrict P.Polarization (span R (range P.root)))) = n := by
+    rw [show (Module.rank R (LinearMap.range (P.Polarization.domRestrict
+      (span R (range P.root))))) = LinearMap.rank (P.Polarization.domRestrict (span R
+      (range P.root))) by rfl, rank_polarization_domRestrict_eq_rank_span_coroot,
+      ← finrank_eq_rank_span_coroot, finrank_span_root_eq]
+  simp [LinearMap.rank_ker_of_nat (LinearMap.domRestrict P.Polarization
+    (span R (range P.root))) hm hn]
+
+-- I would like to use `Submodule.rank_eq_zero` here, but the kernel is a submodule of a submodule.
+lemma polarization_domRestrict_injective :
+    Injective (LinearMap.domRestrict P.Polarization (span R (range P.root))) := by
+  refine LinearMap.ker_eq_bot.mp ?_
+  have : IsReflexive R M := PerfectPairing.reflexive_left P.toPerfectPairing
+  --refine (Submodule.rank_eq_zero (R := R) (M := M)).mp ?_
+  ext x
+  rw [Submodule.mem_bot]
+  refine ⟨fun hx => ?_, fun hx => hx ▸ Submodule.zero_mem _⟩
+  let y : (LinearMap.ker (P.Polarization.domRestrict (span R (range ⇑P.root)))) := ⟨x, hx⟩
+  have : y = 0 := by
+    obtain ⟨a, ⟨h1, h2⟩⟩ := rank_eq_zero_iff.mp (polarization_domRestrict_kernel_rank_zero P) y
+    exact (smul_eq_zero_iff_right h1).mp h2
+  exact (AddSubmonoid.mk_eq_zero _).mp this
```

Comment:
You can proceed with:
```suggestion
  have := P.polarization_domRestrict_kernel_rank_zero
  rwa [Submodule.rank_eq_zero] at this
```
However I think it would be better to proceed as:
```lean
  rw [LinearMap.ker_domRestrict, ← Submodule.disjoint_iff_comap_eq_bot]
```
and restate `polarization_domRestrict_kernel_rank_zero` as suggested there.